polaris-service/src/main/java/org/apache/polaris/service/auth/JWTBroker.java

```diff
@@ -57,35 +58,53 @@ abstract class JWTBroker implements TokenBroker {
   @Override
   public DecodedToken verify(String token) {
     JWTVerifier verifier = JWT.require(getAlgorithm()).build();
-    DecodedJWT decodedJWT = verifier.verify(token);
+    try {
+      DecodedJWT decodedJWT = verifier.verify(token);
+      validateDecodedJWT(decodedJWT);
+      return new DecodedToken() {
+        @Override
+        public Long getPrincipalId() {
+          return decodedJWT.getClaim("principalId").asLong();
+        }
+
+        @Override
+        public String getClientId() {
+          return decodedJWT.getClaim("client_id").asString();
+        }
+
+        @Override
+        public String getSub() {
+          return decodedJWT.getSubject();
+        }
+
+        @Override
+        public String getScope() {
+          return decodedJWT.getClaim("scope").asString();
+        }
+      };
+
+    } catch (JWTVerificationException e) {
+      // Token verification can fail because of following reasons
+      // AlgorithmMismatchException - if the algorithm stated in the token's header it's not
+      // equal to the one defined in the JWTVerifier.
+      // SignatureVerificationException - if the signature is invalid.
+      // TokenExpiredException - if the token has expired.
+      // InvalidClaimException - if a claim contained a different value than the expected one.
+      throw new NotAuthorizedException(
+          "Failed to verify the token with cause %s and reason %s", e.getCause(), e.getMessage());
```

Comment:
Will this message be reported to the client? Exposing too specific validation messages (like Token Expired, Invalid Claim) may not be desirable from the security perspective.

Reply:
I'd suggest logging the specific reason on the server side and returning a generic failure message to the client. WDYT?

If we want debugging aids, we could add the same fresh UUID to both message for correlation.

Reply:
I think "Token expired" should be good others reasons are not required to throw. Also the other [validations](https://github.com/apache/iceberg/blob/5dde680797a21955964298d684f40f97fbc23406/core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java#L586) looks redundant to me. 

```
 if (decodedJWT.getExpiresAtAsInstant().isBefore(Instant.now())) {
      throw new NotAuthorizedException("Token has expired");
    }
```
Adding UUID is a good idea.

Reply:
Yeah, anything that can be derived from the public JWT data (e.g. expiry) is ok to mention in the public error message (to help with trouble-shooting).

Reply:
I've also recently introduced a `PolarisExceptionMapper`, we could use that

Reply:
I have done some modification of token verifying logic. [this](https://github.com/apache/polaris/pull/350/files#diff-5d87d95e854022912a4663deb8d06d209cd84c6e0bafd0a159dc8da2ec24d1ffR60)  should do expiry and token status validation. 
```
JWT.require(getAlgorithm()).withClaim(CLAIM_KEY_ACTIVE, true).build();
```
I have added tests for the same. 

I might need help for adding the trace it. Do we have traceId set in thread local ? or I should just assign a randomId with exception and logs.

Reply:
Although other fix would have been adding `JWTVerificationException` at [IcebergExceptionMapper](https://github.com/apache/polaris/blob/bb4722567846a03aed6c32dec55c4ae0fece9b0c/polaris-service/src/main/java/org/apache/polaris/service/exception/IcebergExceptionMapper.java#L72) as 
```
case JWTVerificationException e -> Response.Status.UNAUTHORIZED.getStatusCode();
```
But this will start showing all the token validation failed reasons  as response entity.

Reply:
@eric-maynard / @collado-mike , PTAL. let me know if there are any thoughts.

Reply:
+1 to adding the traceId to the error message